### PR TITLE
Updated Timeline Semaphore sample

### LINKED ISSUE #588: Timeline semaphore sample not available on Windows
Due to a kernel bug the timeline semaphore sample was disabled on windows:

> Not enabled on Windows at this time due to bugs.
Out-of-order submission in presentation causes kernel level issues,
and need to be figured out before this sample can be enabled on Windows.

We should reevaluate this and if it works on Windows now, the sample should be enabled for that platform again.

### PR DESCRIPTION
**Pull request by bryce-young-mobica**
## Description

Reworked the timeline semaphore sample to prevent it crashing on Windows.

The only obvious trigger for the crash that I could observe was the main thread calling "vkDeviceWaitIdle" whilst the compute thread was in "wait_timeline_gpu". To avoid this, I removed the "wait/signal_timeline_gpu" calls (opting to attach the "VkTimelineSemaphoreSubmitInfo" to the queue submissions instead), and restructured the compute/graphics work stages to prevent the compute thread running ahead and submitting (potentially blocking) work.

- Replaced the "wait/signal_timeline_gpu" calls with "VkTimelineSemaphoreSubmitInfo" added to the queue submissions
- Added an enum for identifying the timeline stages, to make it easier to read and debug the signals/waits
- Moved the rendering work to a seperate thread, leaving the main thread responsible for acquiring and presenting the swapchain image, and signalling the timeline semaphore.

Fixes #588

Tested on Windows
Tested using seperate and shared queue for compute and graphics

## General Checklist:

Please ensure the following points are checked:

- [X] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [X] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [X] I have commented any added functions (in line with Doxygen)
- [X] I have commented any code that could be hard to understand
- [X] My changes do not add any new compiler warnings
- [X] My changes do not add any new validation layer errors or warnings
- [X] I have used existing framework/helper functions where possible
- [X] My changes do not add any regressions
- [X] I have tested every sample to ensure everything runs correctly
- [X] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [X] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [X] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [X] I have tested the sample on at least one compliant Vulkan implementation
- [X] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)
- [X] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [X] Any dependent assets have been merged and published in downstream modules
- [X] For new samples, I have added a paragraph with a summary to the appropriate chapter in the readme of the folder that the sample belongs to [e.g. api samples readme](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/samples/api/README.adoc)
- [X] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/conditional_rendering)
- [X] For new samples, I have added a link to the [Antora navigation](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/antora/modules/ROOT/nav.adoc) so that the sample will be listed at the Vulkan documentation site

### CODE REVIEW

**Review by SaschaWillems — DISMISSED** [review]
Thank you very much for this PR. The sample now finally for me on Windows 11 + RT 4070. No more kernel hang :)

**Review by gary-sweet — DISMISSED** [review]

File: samples/extensions/timeline_semaphore/timeline_semaphore.cpp [hunk 1]
```diff
@@ -17,111 +17,165 @@
...
+namespace
+{
 static constexpr unsigned grid_width  = 64;
 static constexpr unsigned grid_height = 64;
 
-// A simple variant of std::lock_guard which takes a condition for when to lock.
-// We need this since we will only lock vkQueueSubmit when the worker thread needs to submit to the main thread as well.
-// Otherwise, we will submit lock-free since we only need to externally synchronize the same VkQueue.
-class ConditionalLockGuard
+VkTimelineSemaphoreSubmitInfo create_timeline_submit_info(uint32_t waitValueCount, uint64_t *waitValue, uint32_t signalValueCount, uint64_t *signalValue)
```
Comment (asuessenbach) [inline samples/extensions/timeline_semaphore/timeline_semaphore.cpp]: Maybe move this function to common/vk_initializers.h. And rename it to timeline_semaphore_submit_info.

File: samples/extensions/timeline_semaphore/timeline_semaphore.cpp [hunk 2]
```diff
@@ -17,111 +17,165 @@
...
+
+		vkDestroyCommandPool(vk_device, compute.command_pool, nullptr);
+		vkDestroyPipelineLayout(vk_device, compute.pipeline_layout, nullptr);
+		vkDestroyPipeline(vk_device, compute.update_pipeline, nullptr);
+		vkDestroyPipeline(vk_device, compute.mutate_pipeline, nullptr);
+		vkDestroyPipeline(vk_device, compute.init_pipeline, nullptr);
+
+		vkDestroyDescriptorSetLayout(vk_device, shared.storage_layout, nullptr);
+		vkDestroyDescriptorSetLayout(vk_device, shared.sampled_layout, nullptr);
+
```
Comment (asuessenbach) [inline samples/extensions/timeline_semaphore/timeline_semaphore.cpp:67]: You missed to destroy the `shared.descriptor_pool` here.

File: samples/extensions/timeline_semaphore/timeline_semaphore.cpp [hunk 3]
```diff
@@ -17,111 +17,165 @@
...
+		// Need CONCURRENT usage here since we will sample from the image
+		// in both graphics and compute queues.
+		if (get_device().get_queue_family_index(VK_QUEUE_COMPUTE_BIT) !=
+		    get_device().get_queue_by_present(0).get_family_index())
+		{
+			queue_families[0]  = get_device().get_queue_by_present(0).get_family_index();
+			queue_families[1]  = get_device().get_queue_family_index(VK_QUEUE_COMPUTE_BIT);
+			num_queue_families = 2;
+		}
```
Comment (asuessenbach) [inline samples/extensions/timeline_semaphore/timeline_semaphore.cpp:117]: What happens if those two families are equal?

File: samples/extensions/timeline_semaphore/timeline_semaphore.cpp [hunk 4]
```diff
@@ -17,111 +17,165 @@
...
+	// Descriptor sets
+	{
+		VkDescriptorSetAllocateInfo storage_alloc_info = vkb::initializers::descriptor_set_allocate_info(shared.descriptor_pool, &shared.storage_layout, 1);
+		VkDescriptorSetAllocateInfo sampled_alloc_info = vkb::initializers::descriptor_set_allocate_info(shared.descriptor_pool, &shared.sampled_layout, 1);
+
+		for (int i = 0; i < NumAsyncFrames; ++i)
+		{
+			VK_CHECK(vkAllocateDescriptorSets(get_device().get_handle(), &storage_alloc_info, &shared.storage_images[i]));
+			VK_CHECK(vkAllocateDescriptorSets(get_device().get_handle(), &sampled_alloc_info, &shared.sampled_images[i]));
+
```
Comment (asuessenbach) [inline samples/extensions/timeline_semaphore/timeline_semaphore.cpp:161]: Why do you name a `VkDescriptorSet` image?

File: samples/extensions/timeline_semaphore/timeline_semaphore.cpp [hunk 5]
```diff
@@ -17,111 +17,165 @@
...
+			};
+
+			vkUpdateDescriptorSets(get_device().get_handle(), 2, writes, 0, nullptr);
+		}
 	}
 }
 
-void TimelineSemaphore::create_timeline_semaphore(Timeline &timeline)
+void TimelineSemaphore::build_command_buffers()
+{
```
Comment (asuessenbach) [inline samples/extensions/timeline_semaphore/timeline_semaphore.cpp:176]: Empty function?
Maybe add some comment, that this function is needed to resolve the pure virtual from its parent.

File: samples/extensions/timeline_semaphore/timeline_semaphore.cpp [hunk 6]
```diff
@@ -661,54 +586,83 @@ void TimelineSemaphore::render(float delta_time)
...
+	wait_on_timeline(Timeline::present);
+
+	ApiVulkanSample::submit_frame();
+
+	// Signal to the worker threads that they can proceed to the next frame's work
+	signal_next_frame();
 }
 
-void TimelineSemaphore::request_gpu_features(vkb::PhysicalDevice &gpu)
+bool TimelineSemaphore::resize(const uint32_t width, const uint32_t height)
```
Comment (asuessenbach) [inline samples/extensions/timeline_semaphore/timeline_semaphore.cpp]: Is this function still needed?

**Review by asuessenbach — COMMENTED** [review]
Great to have this sample on windows was well!

But I get an error about every second time I start this demo:
`Validation Error: [ UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout ] Object 0: handle = 0x243ed7ef060, type = VK_OBJECT_TYPE_COMMAND_BUFFER; Object 1: handle = 0x944a2c0000000039, type = VK_OBJECT_TYPE_IMAGE; | MessageID = 0x4dae5635 | vkQueueSubmit(): pSubmits[0].pCommandBuffers[0] command buffer VkCommandBuffer 0x243ed7ef060[] expects VkImage 0x944a2c0000000039[] (subresource: aspectMask 0x1 array layer 0, mip level 0) to be in layout VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL--instead, current layout is VK_IMAGE_LAYOUT_UNDEFINED.`

File: samples/extensions/timeline_semaphore/timeline_semaphore.cpp (hunk 1 above)
Comment (bryce-young-mobica) [reply samples/extensions/timeline_semaphore/timeline_semaphore.cpp]: Done.

File: samples/extensions/timeline_semaphore/timeline_semaphore.cpp (hunk 2 above)
Comment (bryce-young-mobica) [reply samples/extensions/timeline_semaphore/timeline_semaphore.cpp:67]: Done

File: samples/extensions/timeline_semaphore/timeline_semaphore.cpp (hunk 5 above)
Comment (bryce-young-mobica) [reply samples/extensions/timeline_semaphore/timeline_semaphore.cpp:176]: Done

File: samples/extensions/timeline_semaphore/timeline_semaphore.cpp (hunk 6 above)
Comment (bryce-young-mobica) [reply samples/extensions/timeline_semaphore/timeline_semaphore.cpp]: Nope, removed it.

File: samples/extensions/timeline_semaphore/timeline_semaphore.cpp (hunk 3 above)
Comment (bryce-young-mobica) [reply samples/extensions/timeline_semaphore/timeline_semaphore.cpp:117]: Not sure, I hadn't tried forcing compute and present to use the same family. 
I've updated this section to properly initialise and set the variables, and only use VK_SHARING_MODE_CONCURRENT if required. 
The app still works with graphics, compute and present using the same queue

**Comment by bryce-young-mobica** [comment]
> But I get an error about every second time I start this demo: `Validation Error: [ UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout ] Object 0: handle = 0x243ed7ef060, type = VK_OBJECT_TYPE_COMMAND_BUFFER; Object 1: handle = 0x944a2c0000000039, type = VK_OBJECT_TYPE_IMAGE; | MessageID = 0x4dae5635 | vkQueueSubmit(): pSubmits[0].pCommandBuffers[0] command buffer VkCommandBuffer 0x243ed7ef060[] expects VkImage 0x944a2c0000000039[] (subresource: aspectMask 0x1 array layer 0, mip level 0) to be in layout VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL--instead, current layout is VK_IMAGE_LAYOUT_UNDEFINED.`

I saw that error during development (iirc it was related to the compute work starting before the setup had finished), and I resolved it by moving the initial "game of life" setup out of the compute thread.

How are you running the tests? (I've been unable to reproduce it myself using the VS debugger or Vulkan Configurator)

**Comment by SaschaWillems** [comment]
I can reproduce that message, but only randomly. I did run the sample many times (latest commit) and only saw that validation error twice. It's triggered from here:

image

Since it's so random (sometimes takes more than 10 runs to show up) it's pretty hard to debug.

P.S. : I'm running a debug build on Win11 with VS 2022 and using the latest SDK.

**Comment by asuessenbach** [comment]
I can repro the validation layer error on about every second run, on Win10, VS2022, NVIDIA RTX A3000 Laptop GPU.

It's always the the `shared.images[0]`, created on line 122.
If I `wait_on_timeline(Timeline::draw);` in `do_graphics_work()` when `timeline.frame == 0`, the issue seems to disappear. But I have no idea, what synchronization is missing or failing to make that a requirement.

**Comment by bryce-young-mobica** [comment]
> It's always the the `shared.images[0]`, created on line 122. If I `wait_on_timeline(Timeline::draw);` in `do_graphics_work()` when `timeline.frame == 0`, the issue seems to disappear. But I have no idea, what synchronization is missing or failing to make that a requirement.

Thanks, I think that is fixing the issue by ensuring the compute commands are submitted first (image[1] is initialised in the "setup_game_of_life" function, but image[0] was only being initialised on that first submission). I've updated the "setup_game_of_life" function to run both VkImages through the compute "init_pipeline". 

@asuessenbach please could you try the latest patch and let me know if it helps?

**Comment by asuessenbach** [comment]
Yep, that seems to fix the issue.
Note, though, that you could handle all NumSyncFrames images in one submit, like so:
[0001-Get-image-initializations-in-one-command-buffer.patch](https://github.com/user-attachments/files/16660000/0001-Get-image-initializations-in-one-command-buffer.patch)

**Review by SaschaWillems — APPROVED** [review]

**Review by asuessenbach — APPROVED** [review]

**Review by gary-sweet — APPROVED** [review]

**Comment by marty-johnson59** [comment]
Merging - 3 approvals

**Pull request merged**